Scripts/Northrend: Necro Overlord Mezhen

TheVanoq opened the pull request:
**Changes proposed:**
Added spells and behavior to "Necro Lord Mezhen":
- He casts a purple beam onto the structure in front of him
- He casts "Raise the Dead", as well as "Drain Soul" when in combat

**Target branch(es):** 3.3.5/master
- 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)
- "Necro Lord Mezhen" had no behavior. He now has an idle cast and the correct combat behavior.

**Tests performed:** (Does it build, tested in-game, etc.)
- Builds and works in-game

**Known issues and TODO list:** (add/remove lines as needed)
- None

TheVanoq committed:
Implemented the behavior of "Necro Lord Mezhen" in "The Howling Fjord".

TheVanoq committed:
Moved from tabs to whitespace

Aokromes commented on sql/updates/world/3.3.5/2017_99_99_99_world.sql:
```diff
@@ -0,0 +1,9 @@
+DELETE FROM conditions WHERE SourceEntry='43559';
+INSERT INTO conditions
+	(SourceTypeOrReferenceId, SourceGroup, SourceEntry, SourceId, ElseGroup, ConditionTypeOrReference, ConditionTarget, ConditionValue1, ConditionValue2, ConditionValue3, NegativeCondition, ErrorType, ErrorTextId, ScriptName, Comment)
+	VALUES (13, 1, 43559, 0, 0, 36, 0, 0, 0, 0, 1, 0, 0, '', '');
```
Don't delete creature template, use update, also missing newline.

Aokromes commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -668,4 +760,5 @@ void AddSC_howling_fjord()
     new npc_riven_widow_cocoon();
- }
+    new npc_necro_overlord_mezhen();
+ }
```
missing newline.

TheVanoq committed:
Added newlines, fixed SQL-Patch

TheVanoq committed:
Notepad++ being dumb and not adding newline?

TheVanoq commented:
Should be all fixed now

Aokromes committed:
Create zone_howling_fjord.cpp

Commit:
SQL code style

- never apostrophes around integer number values
- always back ticks around names of tables and columns
- add comment line on top for TDB content separation

Killyana commented:
It has a quote in aggro:
Necro Overlord Mezhen says: Foolish child, you will not stop what I have set in motion! Now, meet your doom!

Treeston commented:
Refactor using `EventMap` please - see [boss scripts](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp#L144-L185) for reference.

(NB: `BossAI` has a member `EventMap events` that you'll need to declare yourself.)

PS: Code style - data members below methods.

Keader commented:
You can use TaskScheduler. Examples in: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Outland/BlackTemple/boss_illidan.cpp#L1371

TheVanoq committed:
Implemented EventMap

Aokromes commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -15,17 +15,17 @@
-npc_plaguehound_tracker
-npc_apothecary_hanes
-EndContentData */
+ /* ScriptData
```
revert this change

Aokromes commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -35,30 +35,30 @@ EndContentData */
-/*######
-## npc_apothecary_hanes
-######*/
+ /*######
```
revert this change

Aokromes commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -35,30 +35,30 @@ EndContentData */
+ ######*/
 enum Entries
 {
     NPC_APOTHECARY_HANES = 23784,
```
revert this changes

Aokromes commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -394,20 +395,20 @@ class npc_razael_and_lyana : public CreatureScript
 enum Daegarnn
 {
-    QUEST_DEFEAT_AT_RING            = 11300,
+    QUEST_DEFEAT_AT_RING = 11300,
```
revert this change

Aokromes commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -551,61 +552,61 @@ class spell_mindless_abomination_explosion_fx_master : public SpellScriptLoader
-    SPELL_SUMMON_WINTERSKORN_WOODSMAN     = 43283,
-    SPELL_SUMMON_WINTERSKORN_TRIBESMAN    = 43284,
-    SPELL_SUMMON_WINTERSKORN_ORACLE       = 43285,
+    SPELL_SUMMON_BABY_RIVEN_WIDOWS = 43275,
```
revert this changes

tkrokli commented:
Removing the enum alignment completely is an unwanted codestyle change.
Please restore the alignment for the enum values.

Commit:
Add creature_text to SQL file

- insert missing text and values

sirikfoll commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -659,6 +660,111 @@ class npc_riven_widow_cocoon : public CreatureScript
+        void EnterCombat(Unit* /*victim*/) override
+        {
+            events.ScheduleEvent(EVENT_CAST_DRAIN_SOUL, randtime(Seconds(0), Seconds(5)));
+            events.ScheduleEvent(EVENT_CAST_RAISE_DEAD, randtime(Seconds(0), Seconds(10)));
```
you can remove those `randtime`, like this: `events.ScheduleEvent(EVENT_CAST_RAISE_DEAD, Seconds(0), Seconds(10));`

TheVanoq committed:
I hate Visual Studio. Sorry for ruining the formatting.

sirikfoll commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -659,6 +660,111 @@ class npc_riven_widow_cocoon : public CreatureScript
+                            return;
+                        }
+
+                        if (Unit* target = me->FindNearestCreature(24021, 16.0f))
```
And create a enum to npc 24021

tkrokli replied:
Yes, please avoid "magic numbers" as much as possible (use enum).

TheVanoq committed:
Milliseconds() -> Seconds(), Bunny target is now an enum

Keader commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
+
+                        break;
+                    case EVENT_CAST_RAISE_DEAD:
+                        if (!UpdateVictim())
```
it should be on top of this method.

Keader commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
+
+                        DoCast(SPELL_RAISE_DEAD);
+
+                        auto cast_time = Milliseconds(1500);
```
we only use "auto" to iterator

Commit:
Update 2017_99_99_99_world.sql

- correct NPC name in comment (cosmetic)

Keader commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
+                        DoCast(SPELL_RAISE_DEAD);
+
+                        auto cast_time = Milliseconds(1500);
+                        events.Repeat(randtime(cast_time + Seconds(30), cast_time + Seconds(60)));
```
should be: 

``` events.Repeat(Seconds(31)+Milliseconds(500), Seconds(61)+Milliseconds(500));```

Warpten replied:
are you all being serious with your custom std::chrono::fuckwith complaints

Keader commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
+        {
+            events.Reset();
+
+            events.ScheduleEvent(EVENT_CAST_IDLE_BEAM, 333);
```
events.ScheduleEvent(EVENT_CAST_IDLE_BEAM, Milliseconds(333));

Keader commented on src/server/scripts/Northrend/zone_howling_fjord.cpp:
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
+
+                        if (Unit* target = me->FindNearestCreature(BUNNY_BEAM_TARGET, 16.0f))
+                        {
+                            me->CastSpell(target, SPELL_IDLE_PURPLE_BEAM, false);
```
change it to: DoCast(target, SPELL_IDLE_PURPLE_BEAM);

Treeston commented:
@tkrokli please do not make use of "allow edits from maintainers" unless you're in the process of merging a merge-ready PR (to rename SQL files or make very minor changes).

Otherwise, you end up with developers needing to spend valuable time on git merge because somebody felt the need to unilaterally push a whitespace or alignment commit.

tkrokli commented:
OK. Sorry that I misunderstood that feature. But is that the reason this PR was closed? Should a new PR be started instead?

Treeston commented:
Please leave PR management to the person making the PR as well as the reviewing devs - wildly firing off commits often makes the process more complicated than it needs to be for the person contributing.

Killyana commented:
This could be done by SAI:
``` sql
DELETE FROM `creature_text` WHERE `CreatureID` = 24018; -- entire row missing in TDB 335.63
INSERT INTO `creature_text` (`CreatureID`,`GroupID`,`ID`,`Text`,`Type`,`Language`,`Probability`,`Emote`,`Duration`,`Sound`,`BroadcastTextId`,`TextRange`,`comment`) VALUES
(24018, 0, 0, 'Foolish child, you will not stop what I have set in motion! Now, meet your doom!', 12, 0, 100, 0, 0, 0, 23694, 0, 'Necro Overlord Mezhen - SAY_AGGRO');

DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=13 AND `SourceEntry` = 43559;
INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorType`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
(13, 1, 43559, 0, 0, 36, 0, 0, 0, 0, 1, 0, 0, '', '');

DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=22 AND `SourceGroup`=4 AND `SourceEntry` = 24018;
INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorType`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
(22, 4, 24018, 0, 0, 29, 1, 24014, 15, 1, 0, 0, 0, '', '');

UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 24018;
DELETE FROM `smart_scripts` WHERE `entryorguid` IN (24018) AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(24018, 0, 0, 0, 1, 0, 100, 1, 1000, 1000, 0, 0, 11, 43151, 2, 0, 0, 0, 0, 11, 24021, 30, 0, 0, 0, 0, 0, "Necro Overlord Mezhen - Out of Combat - Cast 'Necrolord: Purple Beam'(No Repeat)"),
(24018, 0, 1, 0, 4, 0, 100, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, "Necro Overlord Mezhen - Out aggro - say text"),
(24018, 0, 2, 0, 0, 0, 100, 0, 2000, 4000, 3000, 5000, 11, 32862, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, "Necro Overlord Mezhen - IC - cast Drain soul"),
(24018, 0, 3, 4, 0, 0, 100, 0, 8000, 11000, 15000, 20000, 11, 43559, 2, 0, 0, 0, 0, 19, 24014, 15, 1, 0, 0, 0, 0, "Necro Overlord Mezhen - IC - cast RAISE DEAD"),
(24018, 0, 4, 0, 61, 0, 100, 0, 0, 0, 0, 0, 41, 0, 0, 0, 0, 0, 0, 19, 24014, 15, 1, 0, 0, 0, 0, "Necro Overlord Mezhen - IC - Despawn corpse");
